### ISSUE TITLE
libraries-jitstress failure in System.Diagnostics.Tests.StackFrameTests.Ctor_FNeedFileInfo

### ISSUE THREAD

**Issue by jakobbotsch** [dotnet/runtime#69154]
Seen on testing for #69117.
Pipeline: https://dev.azure.com/dnceng/public/_build/results?buildId=1762407&view=results
Test: https://dev.azure.com/dnceng/public/_build/results?buildId=1762407&view=ms.vss-test-web.build-test-results-tab&runId=47434070&resultId=110968&paneView=debug
Log: https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-69117-merge-a31de8e595db4aa8af/System.Diagnostics.StackTrace.Tests/1/console.b93af7d3.log?helixlogtype=result

```scala
C:\h\w\BB8A09D2\w\AA4A094A\e>set COMPlus 
COMPlus_JitStress=1
COMPlus_TieredCompilation=0

C:\h\w\BB8A09D2\w\AA4A094A\e>call RunTests.cmd --runtime-path C:\h\w\BB8A09D2\p 
----- start Tue 05/10/2022 15:49:12.15 ===============  To repro directly: ===================================================== 
pushd C:\h\w\BB8A09D2\w\AA4A094A\e\
"C:\h\w\BB8A09D2\p\dotnet.exe" exec --runtimeconfig System.Diagnostics.StackTrace.Tests.runtimeconfig.json --depsfile System.Diagnostics.StackTrace.Tests.deps.json xunit.console.dll System.Diagnostics.StackTrace.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing 
popd
===========================================================================================================

C:\h\w\BB8A09D2\w\AA4A094A\e>"C:\h\w\BB8A09D2\p\dotnet.exe" exec --runtimeconfig System.Diagnostics.StackTrace.Tests.runtimeconfig.json --depsfile System.Diagnostics.StackTrace.Tests.deps.json xunit.console.dll System.Diagnostics.StackTrace.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing  
  Discovering: System.Diagnostics.StackTrace.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Diagnostics.StackTrace.Tests (found 53 test cases)
  Starting:    System.Diagnostics.StackTrace.Tests (parallel test collections = on, max threads = 4)
    System.Diagnostics.Tests.StackFrameTests.Ctor_FNeedFileInfo(fNeedFileInfo: False) [FAIL]
      Expected GetILOffset() -1 for InvokeStub_StackFrameTests.Ctor_FNeedFileInfo at offset 44 in file:line:column <filename unknown>:0:0
       to be greater or equal to zero.
      Expected: True
      Actual:   False
      Stack Trace:
        /_/src/libraries/System.Diagnostics.StackTrace/tests/StackFrameTests.cs(171,0): at System.Diagnostics.Tests.StackFrameTests.VerifyStackFrameSkipFrames(StackFrame stackFrame, Boolean isFileConstructor, Int32 skipFrames, MethodInfo expectedMethod, Boolean isCurrentFrame)
           at InvokeStub_StackFrameTests.Ctor_FNeedFileInfo(Object, Object, IntPtr*)
        /_/src/libraries/System.Private.CoreLib/src/System/Reflection/MethodInvoker.cs(60,0): at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
  Finished:    System.Diagnostics.StackTrace.Tests
=== TEST EXECUTION SUMMARY ===
   System.Diagnostics.StackTrace.Tests  Total: 121, Errors: 0, Failed: 1, Skipped: 0, Time: 10.512s
```

**Comment by jakobbotsch** [comment]
@steveharter this does not repro if I revert 5195418.
FWIW, in jitstress mode we make different inlining decisions, so that might be affecting things here.

**Comment by AndyAyersMS** [comment]
> @steveharter this does not repro if I revert [5195418](https://github.com/dotnet/runtime/commit/5195418426468d13b042ae079c65aa05595295b4).
> FWIW, in jitstress mode we make different inlining decisions, so that might be affecting things here.

Right, if you want to depend on certain frames showing up in stack traces make sure those frames involve `NoInline` methods. Otherwise random inlining (via jit stress), profile driven inlining, or changes to the jit can lead to spurious test failures.

**Comment by jkotas** [comment]
@steveharter The reflection invoke PR was reverted in 69373. Please make sure that this and other known stacktrace related issues are addressed before resubmitting.

**Comment by steveharter** [comment]
> FWIW, in jitstress mode we make different inlining decisions, so that might be affecting things here.
>> Right, if you want to depend on certain frames showing up in stack traces make sure those frames involve NoInline methods. Otherwise random inlining (via jit stress), profile driven inlining, or changes to the jit can lead to spurious test failures.

Those tests with `COMPlus_JitStress=1` are missing the stack frame for the actual method invoked, but contains the stack frame for the reflection-created `DynamicMethod` that calls the actual method. Either DynamicMethods are being JITed with `COMPlus_JitStress=1` (which they shouldn't be) or the stack frame isn't there for some reason.

**Comment by jkotas** [comment]
> Those tests with COMPlus_JitStress=1 are missing the stack frame for the actual method invoked, but contains the stack frame for the reflection-created DynamicMethod that calls the actual method.

The JIT is free to inline the actual method invoked into the reflection-created DynamicMethod.

If we want to see the actual method invoked on the stack, we should call it using a function pointer. Calling the actual method invoked using a function pointer would also allow us to reuse the invokers between different methods with the same signature.

**Comment by steveharter** [comment]
> The JIT is free to inline the actual method invoked into the reflection-created DynamicMethod given how things are setup today.

My understanding is that DynamicMethods are never inlined but has been [considered](https://github.com/dotnet/runtime/issues/34500).

FWIW here's a small repro. It fails in a Debug build of the runtime + `COMPlus_JitStress=1`:
```cs
using System.Reflection;
using System.Reflection.Emit;

namespace ConsoleApp
{
    internal class Program
    {
        static void Main(string[] args)
        {

            MethodInfo mi = typeof(Foo).GetMethod(nameof(Foo.CallMe), BindingFlags.Public | BindingFlags.Static)!;
            Action action = CreateInvokeDelegate(mi);

            try
            {
                action();
            }
            catch (Exception e)
            {
                Console.WriteLine(e.ToString());
                Console.WriteLine("Call stack contains CallMe():" + e.ToString().Contains("CallMe()")); // should be True
            }
        }

        static Action CreateInvokeDelegate(MethodInfo method)
        {
            var dm = new DynamicMethod(
                "InvokeStub",
                returnType: null,
                parameterTypes: null,
                restrictedSkipVisibility: true);

            ILGenerator il = dm.GetILGenerator();
            il.Emit(OpCodes.Call, method);
            il.Emit(OpCodes.Ret);

            return (Action)dm.CreateDelegate(typeof(Action), target: null);
        }
    }

    public class Foo
    {
        public static void CallMe()
        {
            throw new Exception("Here");
        }
    }
}
```

**Comment by steveharter** [comment]
Yes using `calli` does work. This is preferred since it is a reasonable change to avoid a breaking change.

```diff
-il.Emit(OpCodes.Call, method);
+il.Emit(OpCodes.Ldftn, method);
+il.EmitCalli(OpCodes.Calli, CallingConventions.Standard, null, null, null);
```

I'll work on doing this along with a hashtable to re-use the invokers.

**Comment by jakobbotsch** [comment]
> ```diff
> -il.Emit(OpCodes.Call, method);
> +il.Emit(OpCodes.Ldftn, method);
> +il.EmitCalli(OpCodes.Calli, CallingConventions.Standard, null, null, null);
> ```

If the plan is to reuse the invoke stubs then I assume in the final implementation the function pointer will be more opaque to the JIT?
Note that we are expecting to do JIT work to optimize patterns like this one and make them inlinable, see #44610.

**Comment by steveharter** [comment]
> Note that we that we are expecting to do JIT work to optimize patterns like this one and make them inlinable

I was wondering if using `calli` to avoid the stack issue will eventually stop working based on future JIT changes. If that's the case, and it isn't desirable or feasible to add something like a `NoTargetInline` decl attribute then we should just declare this as a breaking change for when using Invoke. The stackwalk APIs including `Assembly.GetCallingAssembly()` are already documented to pretty much say don't count on it unless you attribute methods with `MethodImplOptions.NoInlining`; we'd need to expand that to include both direct calls and invoked calls which isn't too much of a stretch IMO.

> If the plan is to reuse the invoke stubs then I assume in the final implementation the function pointer will be more opaque to the JIT?

The way I see it is that `ldftn` won't be used and we need to pass in the pointer directly to the emitted method. Will that prevent any future target inlining via https://github.com/dotnet/runtime/issues/44610? For example:
```cs
        static void Main(string[] args)
        {

            MethodInfo mi = typeof(Foo).GetMethod(nameof(Foo.CallMe), BindingFlags.Public | BindingFlags.Static)!;
            IntPtr methodPtr = mi.MethodHandle.GetFunctionPointer();
            Action<IntPtr> action = CreateInvokeDelegate(mi);

            try
            {
                action(methodPtr);
            }
            catch (Exception e)
            {
                Console.WriteLine(e.ToString());
                Console.WriteLine("Call stack contains CallMe():" + e.ToString().Contains("CallMe()")); // should be True
            }
        }

        static Action<IntPtr> CreateInvokeDelegate(MethodInfo method)
        {
            var dm = new DynamicMethod(
                "InvokeStub",
                returnType: null,
                parameterTypes: new Type[] { typeof(IntPtr) },
                restrictedSkipVisibility: true);

            ILGenerator il = dm.GetILGenerator();
            il.Emit(OpCodes.Ldarg_0);
            il.EmitCalli(OpCodes.Calli, CallingConventions.Standard, typeof(void), Type.EmptyTypes, null);
            il.Emit(OpCodes.Ret);

            return (Action<IntPtr>)dm.CreateDelegate(typeof(Action<IntPtr>), target: null);
        }
```

**Comment by jakobbotsch** [comment]
> The way I see it is that `ldftn` won't be used and we need to pass in the pointer directly to the emitted method. Will that prevent any future target inlining via #44610? For example:

Yes, this should be fine. The only problem I can think of is guarded devirtualization, but we do not ever expect to get PGO data for dynamic methods.

> If that's the case, and it isn't desirable or feasible to add something like a NoTargetInline decl attribute then we should just declare this as a breaking change for when using Invoke.

We also have the `System.StubHelpers.StubHelpers.NextCallReturnAddress` intrinsic. The JIT guarantees that the next IL call will result in a call instruction that is not inlined or tailcalled when this is used. For your case it would be sufficient to call it and discard its result. However, I think given that the function pointer will be opaque to the JIT you should be fine without this.

**Comment by jakobbotsch** [comment]
Fixed by #69575
